modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C

```diff
@@ -0,0 +1,158 @@
+    relative_velocity(0).derivatives() =
+        _secondary_x_dot[_qp].derivatives() - _primary_x_dot[_qp].derivatives();
+    relative_velocity(1).derivatives() =
+        _secondary_y_dot[_qp].derivatives() - _primary_y_dot[_qp].derivatives();
+
+    // Get the component in the tangential direction
+    // TODO: Create a consistent mortar tangential vector field
+    _qp_tangential_velocity = relative_velocity * (_tangents[_qp][0] * _JxW_msm[_qp] * _coord[_qp]);
+  }
+  else
+    _qp_tangential_velocity = std::numeric_limits<Real>::quiet_NaN();
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::computeQpIProperties()
+{
+  // Get the _node_to_weighted_gap map
+  ComputeWeightedGapLMMechanicalContact::computeQpIProperties();
+
+  // Get the _node_to_weighted_tangential_velocity map
+  const auto * const node = _lower_secondary_elem->node_ptr(_i);
+  _node_to_weighted_tangential_velocity[node] += _test[_i][_qp] * _qp_tangential_velocity;
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::residualSetup()
+{
+  // Clear both maps
+  ComputeWeightedGapLMMechanicalContact::residualSetup();
+
+  _node_to_weighted_tangential_velocity.clear();
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::jacobianSetup()
+{
+  residualSetup();
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::post()
+{
+  // Enforce frictional complementarity constraints
+  for (const auto & pr : _node_to_weighted_tangential_velocity)
+  {
+    _weighted_gap_ptr = &_node_to_weighted_gap[pr.first];
+    _tangential_vel_ptr = &pr.second;
+    enforceConstraintOnNode(pr.first);
+  }
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::enforceConstraintOnNode(const Node * const node)
+{
+  ComputeWeightedGapLMMechanicalContact::enforceConstraintOnNode(node);
+
+  // Get friction LM index
+  const auto dof_index = node->dof_number(_sys.number(), _friction_var->number(), 0);
+  ADReal friction_lm_value = _friction_var->getNodalValue(*node);
+  Moose::derivInsert(friction_lm_value.derivatives(), dof_index, 1.);
+  const ADReal & tangential_vel = *_tangential_vel_ptr;
+
+  // Get normal LM index
+  ADReal contact_pressure = _var->getNodalValue(*node);
+  const auto dof_index_cp = node->dof_number(_sys.number(), _var->number(), 0);
+  Moose::derivInsert(contact_pressure.derivatives(), dof_index_cp, 1.);
+
+  const auto & weighted_gap = *_weighted_gap_ptr;
+  ADReal nodal_residual;
+
+  // Primal-dual active set strategy (PDASS)
+  if (contact_pressure < _epsilon || std::isnan(weighted_gap) || std::isnan(tangential_vel))
+    nodal_residual = friction_lm_value;
+  else
+  {
+    auto term_1 = std::max(_mu * (contact_pressure + _c * weighted_gap),
+                           std::abs(friction_lm_value + _c_t * tangential_vel * _dt)) *
+                  friction_lm_value;
+    auto term_2 = _mu * std::max(0.0, contact_pressure + _c * weighted_gap) *
+                  (friction_lm_value + _c_t * tangential_vel * _dt);
+
+    nodal_residual = term_1 - term_2;
```

Comment:
I'm just going to assume that this math is right. You could make `term_1` and `term_2` `const auto` if you make any other changes.

Reply:
I will continue to test this class on new cases. Hopefully we'll catch any issues.